### PR TITLE
Twitch Optout Bot

### PR DESCRIPTION
**Pull request by AlsoGaijutsu**
# Changes

This PR introduces a Cloudflare Worker based Twitch Whispers bot for processing opt-out and opt-in requests. These requests are stored in a Firestore database for later retrieval.

Also introduced is a python based setup script, which is to be run when the worker is deployed for the first time. Instructions for how to setup the worker using this script are in the README.md at the root of optout.

### Opt-out worker

The cloudflare worker is set up to handle the basic Twitch webhook requests (verification and revocation) as well as all whispers sent to the account used in setup.

These requests are verified according to Twitch's specification using `node:crypto`, and will process verified whispers containing `/opt_out` and `/opt_in`. 

These requests are passed along to Firestore using the Firestore REST API, which uses a fresh JWT signed access token for each request.

Finally, a Twitch whisper is sent from the bot account in response. The oauth token generated in setup will expire after 4 hours, so the bot handles Twitch auth errors by regenerating the tokens using the oauth refresh token. This new token pair is pushed as secrets to the bot using the Cloudflare Worker API.

### Opt-out Setup

This script uses the .env configured during the readme to generate an auth URL to retrieve user oauth tokens and then push all secrets to the worker. Once this is done, it creates the Twitch webhook for the specified Cloudflare worker.

## Related Cards

- https://github.com/orgs/vanorsigma/projects/1?pane=issue&itemId=72551263

### CODE REVIEW

File: optout/optout-worker/src/firebase.ts
Comment (jameshi16) [inline optout/optout-worker/src/firebase.ts:1]: Can I confirm that there are no firebase libraries for typescript?

File: optout/optout-worker/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        switch (url.pathname) {
+            case '/discord':
+                return await new Response('Hello Discord!');
+            case '/twitch':
+                return await preHandleTwitch(request, env);
+            default:
+                return new Response('route not found', { status: 404 });
+        }
+    },
+} satisfies ExportedHandler<Env>;
```
Comment (jameshi16) [inline optout/optout-worker/src/index.ts:17]: ?: Why is this line needed?

File: optout/optout-worker/src/twitch.ts [hunk 2]
```diff
@@ -0,0 +1,218 @@
+// TODO: Add proper error handling
```
Comment (jameshi16) [inline optout/optout-worker/src/twitch.ts:1]: Put this out in a card and remove it

File: optout/optout-worker/src/twitch.ts
Comment (jameshi16) [inline optout/optout-worker/src/twitch.ts:1]: To be perfectly honest, I still think this is quite icky. Can we confirm that there are no twitch libraries we can use? (It looks like we're using the proper OAuth workflow here, but I could be wrong)

**Review by jameshi16 — CHANGES_REQUESTED** [review]
some questions

@owobred, do give it a review asw since you should have more context then I do

File: optout/optout-setup/pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,19 @@
+[tool.poetry]
+name = "optout-setup"
+version = "0.1.0"
+description = ""
+authors = ["Gaijutsu", "owobred"]
```
[the remark this answers is not in the document]
Comment (Gaijutsu) [reply optout/optout-setup/pyproject.toml:5]: Looks like it's optional, so I'll just drop it :+1:

File: optout/optout-worker/src/twitch.ts
Comment (Gaijutsu) [reply optout/optout-worker/src/twitch.ts:1]: I'll give libraries another try now that I know how Cloudflare does its node compatibility stuffs :+1:

File: optout/optout-worker/src/firebase.ts
Comment (Gaijutsu) [reply optout/optout-worker/src/firebase.ts:1]: Took another run at it, and  found the `fireworkers` library to work perfectly

File: optout/optout-worker/src/twitch.ts [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+
+    switch (text) {
+        case '/opt_out':
+            await handleOptout(userId, env);
+            return new Response('Success');
+        case '/opt_in':
+            await handleOptin(userId, env);
+            return new Response('Success');
+        default:
+            return new Response('Who asked lol');
```
[the remark this answers is not in the document]
Comment (Gaijutsu) [reply optout/optout-worker/src/twitch.ts:103]: Very good point, I've added some help text to the default text.

**Comment by AlsoGaijutsu** [comment]
Closing as this is now too far out of date

**Pull request closed without merging**